File: array/leetcode/59.go
// 螺旋矩阵2
package leetcode

func generateMatrix(n int) [][]int {
	martix := make([][]int, n)
	for i := range martix {
		martix[i] = make([]int, n)
	}

	top, left, right, bottom := 0, 0, n-1, n-1
	// 	
	num, target := 1, n*n

	for num <= target {
		for i:=left; i <= right; i++ {
			martix[top][i] = num
			num++
		}
		top++
		for i:=top;i<=bottom;i++ {
			martix[i][right] = num
			num++
		}
		right--
		for i:=right;i>=left;i-- {
			martix[bottom][i] = num
			num++
		}
		bottom--
		for i:=bottom;i>=top;i-- {
			martix[i][left] = num
			num++
		}
		left++
	}
	return martix	
}

File: array/cpp/59.cpp
#include <vector>
using namespace std;

class Solution
{
public:
  vector<vector<int>> generateMatrix(int n)
  {
    int top = 0;
    int bottom = n - 1;
    int left = 0;
    int right = n - 1;
    int num = 1;
    vector<vector<int>> res(n, vector<int>(n, 0));
    int tar = n * n;
    while (num <= tar)
    {
      for (int i = left; i <= right; i++)
      {
        res[top][i] = num;
        num++;
      }
      top++;
      for (int i = top; i <= bottom; i++)
      {
        res[i][right] = num;
        num++;
      }
      right--;
      for (int i = right; i >= left; i--)
      {
        res[bottom][right] = num;
        num++;
      }
      bottom--;
      for (int i = bottom; i >= top; i--)
      {
        res[i][left] = num;
        num++;
      }
      left++;
    }
    return res;
  }
};